File: java/src/test/java/co/rerout/sdk/WebhooksTest.java
/*
 * rerout-java — webhook signature verification tests.
 *
 * Copyright (c) 2026 Codecraft Solutions
 * Licensed under the MIT License — https://codecraftsolutions.co.za
 */

package co.rerout.sdk;

import static co.rerout.sdk.Webhooks.verifyReroutSignature;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.nio.charset.StandardCharsets;
import java.util.function.LongSupplier;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

class WebhooksTest {

    private static final String SECRET = "whsec_test_secret";
    private static final String BODY = "{\"event\":\"link.clicked\",\"code\":\"q4\"}";
    private static final long FIXED_NOW = 1_716_000_000L;

    /** Computes a real {@code t=…,v1=…} header for a body under a secret. */
    private static String signHeader(long ts, String signingSecret, String payload) {
        try {
            Mac mac = Mac.getInstance("HmacSHA256");
            mac.init(new SecretKeySpec(
                    signingSecret.getBytes(StandardCharsets.UTF_8), "HmacSHA256"));
            byte[] raw = mac.doFinal((ts + "." + payload).getBytes(StandardCharsets.UTF_8));
            StringBuilder hex = new StringBuilder(raw.length * 2);
            for (byte b : raw) {
                hex.append(Character.forDigit((b >> 4) & 0xF, 16));
                hex.append(Character.forDigit(b & 0xF, 16));
            }
            return "t=" + ts + ",v1=" + hex;
        } catch (java.security.GeneralSecurityException e) {
            throw new IllegalStateException(e);
        }
    }

    private static String signHeader() {
        return signHeader(FIXED_NOW, SECRET, BODY);
    }

    private static LongSupplier clock(long at) {
        return () -> at;
    }

    private static LongSupplier clock() {
        return clock(FIXED_NOW);
    }

    // ─── valid ──────────────────────────────────────────────────────────────

    @Test
    void aFreshlySignedPayloadVerifies() {
        assertTrue(verifyReroutSignature(
                BODY, signHeader(), SECRET, Webhooks.DEFAULT_TOLERANCE_SECONDS, clock()));
    }

    @Test
    void theDefaultToleranceConvenienceFormVerifies() {
        // System clock — sign at "now" so it falls inside the 300s window.
        long now = System.currentTimeMillis() / 1000L;
        assertTrue(verifyReroutSignature(
                BODY, signHeader(now, SECRET, BODY), SECRET));
    }

    // ─── wrong / tampered ───────────────────────────────────────────────────

    @Test
    void aSignatureMadeWithADifferentSecretIsRejected() {
        String header = signHeader(FIXED_NOW, "whsec_other", BODY);
        assertFalse(verifyReroutSignature(
                BODY, header, SECRET, Webhooks.DEFAULT_TOLERANCE_SECONDS, clock()));
    }

    @Test
    void aTamperedBodyWithAnExtraSpaceIsRejected() {
        String header = signHeader();
        assertFalse(verifyReroutSignature(
                BODY + " ", header, SECRET, Webhooks.DEFAULT_TOLERANCE_SECONDS, clock()));
    }

    @Test
    void aBodyWithAMissingCharacterIsRejected() {
        String header = signHeader();
        assertFalse(verifyReroutSignature(
                BODY.substring(0, BODY.length() - 1), header, SECRET,
                Webhooks.DEFAULT_TOLERANCE_SECONDS, clock()));
    }

    // ─── timestamp tolerance ────────────────────────────────────────────────

    @Test
    void aPayloadOlderThanTheToleranceWindowIsRejected() {
        String header = signHeader(FIXED_NOW - 600, SECRET, BODY);
        assertFalse(verifyReroutSignature(
                BODY, header, SECRET, Webhooks.DEFAULT_TOLERANCE_SECONDS, clock()));
    }

    @Test
    void aPayloadFromTheFutureBeyondToleranceIsRejected() {
        String header = signHeader(FIXED_NOW + 600, SECRET, BODY);
        assertFalse(verifyReroutSignature(
                BODY, header, SECRET, Webhooks.DEFAULT_TOLERANCE_SECONDS, clock()));
    }

    @Test
    void aPayloadExactlyAtTheToleranceBoundaryIsAccepted() {
        String header = signHeader(FIXED_NOW - 300, SECRET, BODY);
        assertTrue(verifyReroutSignature(
                BODY, header, SECRET, Webhooks.DEFAULT_TOLERANCE_SECONDS, clock()));
    }

    @Test
    void aPayloadOneSecondPastTheBoundaryIsRejected() {
        String header = signHeader(FIXED_NOW - 301, SECRET, BODY);
        assertFalse(verifyReroutSignature(
                BODY, header, SECRET, Webhooks.DEFAULT_TOLERANCE_SECONDS, clock()));
    }

    @Test
    void toleranceZeroDisablesTheTimestampCheck() {
        String header = signHeader(FIXED_NOW - 100_000, SECRET, BODY);
        assertTrue(verifyReroutSignature(BODY, header, SECRET, 0, clock()));
    }

    @Test
    void aCustomToleranceWindowIsHonoured() {
        String header = signHeader(FIXED_NOW - 50, SECRET, BODY);
        assertTrue(verifyReroutSignature(BODY, header, SECRET, 60, clock()));
        assertFalse(verifyReroutSignature(BODY, header, SECRET, 30, clock()));
    }

    // ─── malformed headers ──────────────────────────────────────────────────

    @ParameterizedTest
    @ValueSource(strings = {
        "",
        "garbage",
        "t=,v1=abc123",
        "v1=abc123",
        "t=1716000000",
        "t=notanumber,v1=abc123",
        "t=1716000000,v1=zzzz",
        "t=1716000000,v1=abc",
        "t=-5,v1=abcd",
        "t=0,v1=abcd",
    })
    void malformedHeadersAreRejected(String header) {
        assertFalse(verifyReroutSignature(BODY, header, SECRET, 0, clock()));
    }

    @Test
    void aNullHeaderIsRejected() {
        assertFalse(verifyReroutSignature(
                BODY, null, SECRET, Webhooks.DEFAULT_TOLERANCE_SECONDS, clock()));
    }

    @Test
    void anEmptyHeaderIsRejected() {
        assertFalse(verifyReroutSignature(
                BODY, "", SECRET, Webhooks.DEFAULT_TOLERANCE_SECONDS, clock()));
    }

    // ─── header casing ──────────────────────────────────────────────────────

    @Test
    void uppercaseHeaderKeysAreAccepted() {
        String upper = signHeader().replace("t=", "T=").replace("v1=", "V1=");
        assertTrue(verifyReroutSignature(
                BODY, upper, SECRET, Webhooks.DEFAULT_TOLERANCE_SECONDS, clock()));
    }

    @Test
    void mixedCaseAndPaddedHeaderKeysAreAccepted() {
        String[] parts = signHeader().split(",");
        String padded = " T = " + parts[0].substring(parts[0].indexOf('=') + 1)
                + " , V1 = " + parts[1].substring(parts[1].indexOf('=') + 1) + " ";
        assertTrue(verifyReroutSignature(
                BODY, padded, SECRET, Webhooks.DEFAULT_TOLERANCE_SECONDS, clock()));
    }

    // ─── secret ─────────────────────────────────────────────────────────────

    @Test
    void anEmptySecretIsRejected() {
        assertFalse(verifyReroutSignature(
                BODY, signHeader(), "", Webhooks.DEFAULT_TOLERANCE_SECONDS, clock()));
    }

    @Test
    void aNullSecretIsRejected() {
        assertFalse(verifyReroutSignature(
                BODY, signHeader(), null, Webhooks.DEFAULT_TOLERANCE_SECONDS, clock()));
    }

    // ─── extra segments ─────────────────────────────────────────────────────

    @Test
    void unknownExtraSegmentsInTheHeaderAreIgnored() {
        String header = signHeader() + ",v0=legacy,foo=bar";
        assertTrue(verifyReroutSignature(
                BODY, header, SECRET, Webhooks.DEFAULT_TOLERANCE_SECONDS, clock()));
    }

    @Test
    void theDefaultToleranceIsFiveMinutes() {
        assertEquals(300, Webhooks.DEFAULT_TOLERANCE_SECONDS);
    }
}

File: kotlin/src/test/kotlin/co/rerout/sdk/WebhooksTest.kt
/*
 * rerout-kotlin — webhook signature verification tests.
 *
 * Copyright (c) 2026 Codecraft Solutions
 * Licensed under the MIT License — https://codecraftsolutions.co.za
 */

package co.rerout.sdk

import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.ValueSource
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

class WebhooksTest {
    private val secret = "whsec_test_secret"
    private val body = """{"event":"link.clicked","code":"q4"}"""
    private val fixedNow = 1_716_000_000L

    /** Compute a real `t=…,v1=…` header for [body] under [secret] at [ts]. */
    private fun signHeader(
        ts: Long = fixedNow,
        signingSecret: String = secret,
        payload: String = body,
    ): String {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(signingSecret.toByteArray(Charsets.UTF_8), "HmacSHA256"))
        val hex = mac.doFinal("$ts.$payload".toByteArray(Charsets.UTF_8))
            .joinToString("") { "%02x".format(it) }
        return "t=$ts,v1=$hex"
    }

    private fun clock(at: Long = fixedNow): () -> Long = { at }

    // ─── valid ──────────────────────────────────────────────────────────────

    @Test
    fun `a freshly signed payload verifies`() {
        assertTrue(
            verifyReroutSignature(body, signHeader(), secret, now = clock()),
        )
    }

    @Test
    fun `the object form verifies a fresh payload`() {
        assertTrue(
            ReroutWebhooks.verifySignature(body, signHeader(), secret, now = clock()),
        )
    }

    // ─── wrong / tampered ───────────────────────────────────────────────────

    @Test
    fun `a signature made with a different secret is rejected`() {
        val header = signHeader(signingSecret = "whsec_other")
        assertFalse(verifyReroutSignature(body, header, secret, now = clock()))
    }

    @Test
    fun `a tampered body is rejected`() {
        val header = signHeader()
        assertFalse(verifyReroutSignature("$body ", header, secret, now = clock()))
    }

    @Test
    fun `a body with a missing character is rejected`() {
        val header = signHeader()
        assertFalse(verifyReroutSignature(body.dropLast(1), header, secret, now = clock()))
    }

    // ─── timestamp tolerance ────────────────────────────────────────────────

    @Test
    fun `a payload older than the tolerance window is rejected`() {
        val header = signHeader(ts = fixedNow - 600)
        assertFalse(verifyReroutSignature(body, header, secret, now = clock()))
    }

    @Test
    fun `a payload from the future beyond tolerance is rejected`() {
        val header = signHeader(ts = fixedNow + 600)
        assertFalse(verifyReroutSignature(body, header, secret, now = clock()))
    }

    @Test
    fun `a payload exactly at the tolerance boundary is accepted`() {
        val header = signHeader(ts = fixedNow - 300)
        assertTrue(verifyReroutSignature(body, header, secret, now = clock()))
    }

    @Test
    fun `a payload one second past the boundary is rejected`() {
        val header = signHeader(ts = fixedNow - 301)
        assertFalse(verifyReroutSignature(body, header, secret, now = clock()))
    }

    @Test
    fun `tolerance zero disables the timestamp check`() {
        val header = signHeader(ts = fixedNow - 100_000)
        assertTrue(
            verifyReroutSignature(body, header, secret, toleranceSeconds = 0, now = clock()),
        )
    }

    @Test
    fun `a custom tolerance window is honoured`() {
        val header = signHeader(ts = fixedNow - 50)
        assertTrue(
            verifyReroutSignature(body, header, secret, toleranceSeconds = 60, now = clock()),
        )
        assertFalse(
            verifyReroutSignature(body, header, secret, toleranceSeconds = 30, now = clock()),
        )
    }

    // ─── malformed headers ──────────────────────────────────────────────────

    @ParameterizedTest
    @ValueSource(
        strings = [
            "",
            "garbage",
            "t=,v1=abc123",
            "v1=abc123",
            "t=1716000000",
            "t=notanumber,v1=abc123",
            "t=1716000000,v1=zzzz",
            "t=1716000000,v1=abc",
            "t=-5,v1=abcd",
            "t=0,v1=abcd",
        ],
    )
    fun `malformed headers are rejected`(header: String) {
        assertFalse(
            verifyReroutSignature(body, header, secret, toleranceSeconds = 0, now = clock()),
        )
    }

    @Test
    fun `a null header is rejected`() {
        assertFalse(verifyReroutSignature(body, null, secret, now = clock()))
    }

    // ─── header casing ──────────────────────────────────────────────────────

    @Test
    fun `uppercase header keys are accepted`() {
        val signed = signHeader()
        val upper = signed.replace("t=", "T=").replace("v1=", "V1=")
        assertTrue(verifyReroutSignature(body, upper, secret, now = clock()))
    }

    @Test
    fun `mixed-case and padded header keys are accepted`() {
        val signed = signHeader()
        val parts = signed.split(",")
        val padded = " T = ${parts[0].substringAfter("=")} , V1 = ${parts[1].substringAfter("=")} "
        assertTrue(verifyReroutSignature(body, padded, secret, now = clock()))
    }

    // ─── secret ─────────────────────────────────────────────────────────────

    @Test
    fun `an empty secret is rejected`() {
        assertFalse(verifyReroutSignature(body, signHeader(), "", now = clock()))
    }

    @Test
    fun `a null secret is rejected`() {
        assertFalse(verifyReroutSignature(body, signHeader(), null, now = clock()))
    }

    // ─── extra segments ─────────────────────────────────────────────────────

    @Test
    fun `unknown extra segments in the header are ignored`() {
        val header = "${signHeader()},v0=legacy,foo=bar"
        assertTrue(verifyReroutSignature(body, header, secret, now = clock()))
    }

    @Test
    fun `the default tolerance is five minutes`() {
        assertTrue(ReroutWebhooks.DEFAULT_TOLERANCE_SECONDS == 300L)
    }
}
